src/workers/jobs.js

```diff
@@ -183,55 +224,137 @@ export async function loadContactsFromDataWarehouse(job) {
     return
   }
 
-  // TODO break up result and save in portions, dispatched
   const jobMessages = []
   const savePortion = await Promise.all(knexResult.rows.map(async (row) => {
     const contact = {
-      campaign_id: job.campaign_id,
-      first_name: row.first_name,
-      last_name: row.last_name,
+      campaign_id: jobEvent.campaignId,
+      first_name: row.first_name || '',
+      last_name: row.last_name || '',
       cell: row.cell,
-      zip: row.zip,
-      external_id: row.external_id
+      zip: row.zip || '',
+      external_id: (row.external_id ? String(row.external_id) : ''),
+      assignment_id: null,
+      message_status: 'needsMessage'
     }
     const contactCustomFields = {}
-    for (const f in customFields) {
+    Object.keys(customFields).forEach((f) => {
       contactCustomFields[f] = row[f]
-    }
+    })
     contact.custom_fields = JSON.stringify(contactCustomFields)
     if (contact.zip) {
       contact.timezone_offset = await getTimezoneByZip(contact.zip)
     }
     return contact
   }))
 
-  await r.table('campaign_contact')
-    .getAll(job.campaign_id, { index: 'campaign_id' })
-    .delete()
-
   await CampaignContact.save(savePortion)
 
-  // now that we've saved them all, we delete everyone that is opted out locally
-  // doing this in one go so that we can get the DB to do the indexed cell matching
+  await r.knex('job_request').where('id', jobEvent.jobId).increment('status', 1)
+
+  const completed = (await r.knex('job_request')
+                     .where('id', jobEvent.jobId)
+                     .select('status')
+                     .first())
+  console.log('loadContactsFromDataWarehouseFragment toward end', completed, jobEvent)
+
+  if (!completed) {
+    console.log('loadContactsFromDataWarehouseFragment job has been deleted', completed)
+  } else if (jobEvent.totalParts && completed.status >= jobEvent.totalParts) {
+    if (jobEvent.organizationId) {
+      // now that we've saved them all, we delete everyone that is opted out locally
+      // doing this in one go so that we can get the DB to do the indexed cell matching
+      const optOutCellCount = await r.knex('campaign_contact')
+        .whereIn('cell', function optouts() {
+          this.select('cell').from('opt_out').where('organization_id', jobEvent.organizationId)
+        })
+        .where('campaign_id', jobEvent.campaignId)
+        .delete()
+      console.log('OPTOUT CELL COUNT', optOutCellCount)
+    }
+    await r.table('job_request').get(jobEvent.jobId).delete()
+    return { 'completed': 1 }
+  } else if (jobEvent.part < (jobEvent.totalParts - 1)) {
```

Comment:
lots of complexity in here, and when we have a reliable unit test framework that can setup and tear down dbs without race conditions, i'd ask that you create unit tests for cases: single batch < 10K in size, a batch > 10K, a batch of exactly 30K (to check for off by one errors). For now, could you create unit test stubs with descriptive names so we dont forget to unit test this later?